src/main/scala/lms/collection/ArrayOps.scala

```diff
@@ -8,7 +8,27 @@ import lms.core.virtualize
 import lms.core.utils.time
 import lms.macros.SourceContext
 
-trait ArrayOps { b: Base =>
+trait ArrayOps extends Base with PrimitiveOps {
+
+  case class ARRAY(x: Backend.Exp) {
+    def withSource(pos: SourceContext) = { Adapter.sourceMap(x) = pos; this }
+    def withEleType(m: Manifest[_]) = { Adapter.typeMap(x) = m; this }
+    def with_(pos: SourceContext, m: Manifest[_]) = withSource(pos).withEleType(m)
+
+    def p: SourceContext = Adapter.sourceMap.getOrElse(x, ???)
```

Comment:
Let's only use `???` for work in progress code, it is not intended to use `???` as a general exception. If there is a case that the program should abort, let it die with proper and meaningful error messages. In this case, if there is no default value at all, we can simply write `Adapter.sourceMap(x)`.